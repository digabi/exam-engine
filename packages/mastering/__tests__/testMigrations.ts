import { parseExam, migrateExam } from '@digabi/exam-engine-mastering'
import { readFixture } from './fixtures'
import { wrap } from 'jest-snapshot-serializer-raw'

describe('Exam migrations', () => {
  it('migrates an old exam to the current schema', async () => {
    const xml = await readFixture('old_schema_version.xml')
    const doc = parseExam(xml, false)
    migrateExam(doc)

    expect(wrap(doc.toString(false))).toMatchSnapshot()
  })

  it('does nothing if the exam already is in the latest schema', async () => {
    const xml = await readFixture('minimal_yo_exam.xml')
    const doc = parseExam(xml)
    migrateExam(doc)

    expect(doc.toString(false)).toEqual(parseExam(xml).toString(false))
  })

  it('throws an error if schema version is invalid', async () => {
    const xml = await readFixture('invalid_schema_version.xml')
    const doc = parseExam(xml)

    expect(() => migrateExam(doc)).toThrow('Unsupported exam schema version: 2.0')
  })

  it('throws an error if schema version is missing', async () => {
    const xml = await readFixture('no_schema_version.xml')
    const doc = parseExam(xml)

    expect(() => migrateExam(doc)).toThrow('No exam-schema-version attribute found')
  })
})
