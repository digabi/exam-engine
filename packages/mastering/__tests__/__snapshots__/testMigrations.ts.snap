// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exam migrations migrates an old exam to the current schema 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<e:exam xmlns:e="http://ylioppilastutkinto.fi/exam.xsd" xmlns="http://www.w3.org/1999/xhtml" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://ylioppilastutkinto.fi/exam.xsd https://abitti.dev/schema/exam.xsd" exam-schema-version="0.3" date="2020-01-01" exam-code="A" day-code="X">
    <e:languages>
        <e:language>fi-FI</e:language>
        <e:language>sv-FI</e:language>
    </e:languages>
    <e:exam-title/>
    <e:exam-instruction/>
    <e:section>
        <e:section-title/>
        <e:question>
            <e:question-title/>
            <e:audio src="1.ogg"><e:audio-title>
                Audio file <em>description</em>
            </e:audio-title></e:audio>
            <e:image src="1.jpg"><e:image-title>
                Audio file <em>description</em>
            </e:image-title></e:image>
        </e:question>
    </e:section>
</e:exam>

`;
